packages/flutter/lib/src/widgets/asset_vendor.dart

```diff
@@ -0,0 +1,57 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/services.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+
+/// Establishes an asset resolution strategy for its descendants.
+///
+/// An AssetVendor establishes a [ResolvingAssetBundle] as the default bundle
+/// for its subtree. The resolution logic is provided by an [AssetResolver]
+/// specified via its [AssetResolverBuilder].
+class AssetVendor extends StatefulComponent {
```

Comment:
I'd just fold this class into `ResolutionAwareAssetVendor`.  Having it separate is just making things more complicated.  When we have more than one type of asset vendor, it will probably be clearer how to extract common functionality.

By merging it into ResolutionAwareAssetVendor, you get away from all the complexity around figuring out when to discard your resolve and your cached bundle.  All you need to do is see if the devicePixelRatio in your config changed.

It's likely we'll end up somewhere close to what you've got in this patch, but at the moment it's got more machinery than we need.  It's easy to introduce complexity later, but it's hard to remove it once you've got it.